### PR TITLE
fix(profiling): release the EventPipe session when the SDK shuts down

### LINKED ISSUE #5418: SamplingTransactionProfilerFactory.Dispose disposes the Task, not the SampleProfilerSession (resource leak)
## Summary

`SamplingTransactionProfilerFactory.Dispose()` contains a bug where the `SampleProfilerSession` is never actually disposed, leaving the underlying `EventPipeSession` open.

## Code

**File:** `src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs`

```csharp
public void Dispose()
{
    _sessionTask.ContinueWith(session => session.Dispose());
}
```

`_sessionTask` is `Task<SampleProfilerSession>`. The continuation parameter `session` is therefore the **antecedent `Task<SampleProfilerSession>`**, not the `SampleProfilerSession` result it wraps. Calling `session.Dispose()` disposes the `Task` object, not the session.

## Impact

`SampleProfilerSession.Dispose()` is never called:

```csharp
// SampleProfilerSession.cs
public void Dispose() => Stop();

private void Stop()
{
    // ...
    _session.Stop();      // EventPipeSession — holds a pipe connection to the .NET runtime
    _session.Dispose();
    EventSource.Dispose();
}
```

The `EventPipeSession` holds a connection to the .NET runtime's diagnostics EventPipe. When the profiler factory is disposed (e.g., during SDK shutdown), this connection is left open, causing a resource leak. This is consistent with the Windows Service memory growth reported in #3375, which was linked to Tracing and Profiling being enabled.

## Suggested Fix

```csharp
public void Dispose()
{
    _sessionTask.ContinueWith(
        t => t.Result.Dispose(),
        TaskContinuationOptions.OnlyOnRanToCompletion);
}
```

This ensures the `SampleProfilerSession` is disposed only when the task completed successfully (i.e., a session was actually started), which also avoids accessing `Result` on a faulted or cancelled task.

### PR DESCRIPTION
**Pull request by jamescrosswell**
`SamplingTransactionProfilerFactory.Dispose()` disposed the antecedent `Task<SampleProfilerSession>` rather than the session it wraps, so the `EventPipeSession` was never released.

That line turned out to be the last of three independent defects sitting between "SDK shuts down" and "EventPipe session released" — fixing it alone would not have changed anything observable.

### What was broken

1. **`SampleProfilerSession.Stop()` skipped its own disposals on the happy path.** `_processing` was built with `TaskContinuationOptions.OnlyOnFaulted`. When `eventSource.Process()` returns normally the continuation criteria aren't met, so the continuation transitions to `Canceled` and `_processing.Wait()` throws `AggregateException(TaskCanceledException)`. `EventPipeSession.Dispose()` and `TraceLogEventSource.Dispose()` sat after that call and were never reached — the `catch` logged `Error during sampler profiler session shutdown.` and swallowed it. This fired on every clean shutdown, not on some edge case.

2. **Nothing ever called `Dispose()` on the factory.** `ProfilingIntegration` wasn't `IDisposable`, so `Hub` never added it to `_integrationsToCleanup`, and the factory's only other reference is `options.TransactionProfilerFactory`, which nothing disposes. Outside of tests, `SamplingTransactionProfilerFactory.Dispose()` was unreachable.

3. **The reported bug.** The `ContinueWith` parameter is the antecedent task, so `Dispose()` called `Task.Dispose()` — a near no-op on a completed task — instead of disposing the session.

### What this changes

**`ProfilingIntegration` now owns the factory it creates.** It implements `IDisposable` so the Hub tears it down on shutdown, and disposal also clears `options.TransactionProfilerFactory`, so a `SentryOptions` reused for a later hub gets a working profiler rather than a stale reference. It only touches a factory it created itself — one supplied from elsewhere is left alone.

**`SamplingTransactionProfilerFactory.Dispose()` now actually stops the session.** `StartEventPipeSession()` can block for up to 30 seconds and can't be cancelled, so disposal can land before there is any session to stop; a lock hands the session between the startup task and `Dispose()` so exactly one of the two tears it down. The wait for the first event is cancelled on disposal rather than left to block indefinitely. Both waits are bounded at 2 seconds so a wedged session can't hang shutdown.

**`SampleProfilerSession.Stop()` moves the disposals into a `finally`,** so the EventPipe connection is released even when stopping the session or draining its events fails.

### Impact

Low, and narrower than the issue suggests. The session is created once and lives for the SDK's lifetime — `SamplingTransactionProfiler.Stop()` only clears `_inProgress`, it never touches the session — so the only disposal point is SDK shutdown, which is normally process shutdown, where the OS reclaims the handle and the `TraceLog` regardless. Nothing accumulates while an application runs.

It matters where the process outlives the SDK: repeated `Init`/`Close` cycles strand one `EventPipeSession` plus its `TraceLog` per cycle, and hosts that close Sentry but keep running never release the pipe.

The issue links this to the Windows Service memory growth in #3375. That link doesn't hold up — a service that initialises Sentry once would never reach any of these paths. The steady-state growth reported in #5469 is a separate mechanism and is not addressed here.

### Notes

- Defect 1 has a visible signature in diagnostic logs: `Error during sampler profiler session shutdown.` on every clean shutdown with profiling enabled. Worth asking anyone reporting profiler-related resource growth whether they see it.
- `Hub.Dispose()` now does real work for profiling users where it previously did none.
- Most of `Sentry.Profiling.Tests` is `Skip.If(TestEnvironment.IsGitHubActions)` because it opens real EventPipe sessions. The `ProfilingIntegration` tests added here are plain `[Fact]`s that do run in CI; the two session-level tests follow the existing skip pattern.
- `AGENTS.md` gains a short section recording the repo's preference for code over comments. Happy to split that into its own PR if it doesn't belong here.

Fixes #5418

### CODE REVIEW

File: src/Sentry.Profiling/SampleProfilerSession.cs [hunk 1]
```diff
@@ -47,6 +47,10 @@ private SampleProfilerSession(SentryStopwatch stopwatch, EventPipeSession sessio
     // need a large buffer if we're connecting righ away. Leaving it too large increases app memory usage.
     internal static int CircularBufferMB = 16;
 
+    // How long Stop() waits for the event processing task to drain after the session has been stopped.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SampleProfilerSession.cs]:
```suggestion
```

File: src/Sentry.Profiling/SampleProfilerSession.cs [hunk 2]
```diff
@@ -119,19 +126,39 @@ public async Task WaitForFirstEventAsync(CancellationToken cancellationToken = d
...
+            return;
+        }
+
+        _stopped = true;
+        try
+        {
+            _session.Stop();
+
+            // Let the processing task drain the events that are still in flight, but don't hold up
+            // shutdown indefinitely if it doesn't get there.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SampleProfilerSession.cs]:
```suggestion
```

File: src/Sentry.Profiling/SampleProfilerSession.cs [hunk 3]
```diff
@@ -119,19 +126,39 @@ public async Task WaitForFirstEventAsync(CancellationToken cancellationToken = d
...
+            }
+        }
+        catch (Exception ex)
+        {
+            _logger?.LogWarning(ex, "Error during sampler profiler session shutdown.");
+        }
+        finally
         {
+            // These need to happen even if stopping the session or draining the events failed, otherwise
+            // the EventPipe connection to the runtime is left open.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SampleProfilerSession.cs]:
```suggestion
            // These always need to happen, otherwise the EventPipe connection to the 
            // runtime is left open.
```

File: src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs [hunk 4]
```diff
@@ -14,10 +14,26 @@ internal class SamplingTransactionProfilerFactory : IDisposable, ITransactionPro
     // Stop profiling after the given number of milliseconds.
     private const int TIME_LIMIT_MS = 30_000;
 
+    // How long Dispose() waits for an in-flight session startup to complete before giving up on it.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs]:
```suggestion
    // How long Dispose() waits for an in-flight session to complete before giving up on it.
```

File: src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs [hunk 5]
```diff
@@ -14,10 +14,26 @@ internal class SamplingTransactionProfilerFactory : IDisposable, ITransactionPro
...
+    private const int SHUTDOWN_TIMEOUT_MS = 2_000;
+
     private readonly SentryOptions _options;
 
     internal Task<SampleProfilerSession> _sessionTask;
 
+    // Cancels the wait for the first event so that Dispose() doesn't have to wait for a session that
+    // may never receive one.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs]:
```suggestion
```

File: src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs [hunk 6]
```diff
@@ -14,10 +14,26 @@ internal class SamplingTransactionProfilerFactory : IDisposable, ITransactionPro
...
+    private readonly CancellationTokenSource _shutdownCts = new();
+
+    // Assigned as soon as the session exists, which is earlier than _sessionTask completing. Dispose()
+    // uses this so it can also stop a session that never saw its first event.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs]:
```suggestion
```

File: src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs [hunk 7]
```diff
@@ -83,6 +100,34 @@ public SamplingTransactionProfilerFactory(SentryOptions options, TimeSpan startu
...
+            return;
+        }
+
+        // Unblocks the startup task if it's still waiting for the first event to arrive.
+        _shutdownCts.Cancel();
+
+        try
+        {
+            // Gives an in-flight startup a chance to finish, and observes the exception if it failed
+            // or was cancelled above.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs]:
```suggestion
```

File: src/Sentry.Profiling/SampleProfilerSession.cs [hunk 8]
```diff
@@ -56,6 +59,15 @@ private SampleProfilerSession(SentryStopwatch stopwatch, EventPipeSession sessio
 
     public TraceLog TraceLog => EventSource.TraceLog;
 
+    // Exposed for tests.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SampleProfilerSession.cs]:
```suggestion
```

File: src/Sentry.Profiling/SampleProfilerSession.cs [hunk 9]
```diff
@@ -56,6 +59,15 @@ private SampleProfilerSession(SentryStopwatch stopwatch, EventPipeSession sessio
...
+    internal bool IsStopped => _stopped;
+
+    // Exposed for tests: stands in for the uncancellable StartEventPipeSession() window.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SampleProfilerSession.cs]:
```suggestion
```

File: src/Sentry.Profiling/SampleProfilerSession.cs [hunk 10]
```diff
@@ -56,6 +59,15 @@ private SampleProfilerSession(SentryStopwatch stopwatch, EventPipeSession sessio
...
+    internal static Action? BeforeStartupForTests;
+
+    // Exposed for tests. A callback rather than a static field, so no session is held in production.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SampleProfilerSession.cs]:
```suggestion
```

File: src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs [hunk 11]
```diff
@@ -14,23 +14,60 @@ internal class SamplingTransactionProfilerFactory : IDisposable, ITransactionPro
...
 
+    private readonly CancellationTokenSource _shutdownCts = new();
+
+    // StartNew() can block for up to 30s uncancellably, so Dispose() may run before there is a session
+    // to stop. This hands it between the two sides so exactly one of them stops it.
+    private readonly object _sessionLock = new();
+    private SampleProfilerSession? _session;
+    private bool _disposed;
+
+    // Exposed for tests.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs]:
```suggestion
```

File: src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs [hunk 12]
```diff
@@ -14,23 +14,60 @@ internal class SamplingTransactionProfilerFactory : IDisposable, ITransactionPro
...
+        get { lock (_sessionLock) { return _disposed; } }
+    }
+
     private bool _errorLogged = false;
 
     public SamplingTransactionProfilerFactory(SentryOptions options, TimeSpan startupTimeout)
     {
         _options = options;
 
+        // Reading _shutdownCts.Token once Dispose() has disposed it throws.
```
Comment (jamescrosswell) [inline src/Sentry.Profiling/SamplingTransactionProfilerFactory.cs]:
```suggestion
        // Store local reference to avoid ObjectDisposed exception
```

File: test/Sentry.Profiling.Tests/ProfilingSentryOptionsExtensionsTests.cs [hunk 13]
```diff
@@ -61,6 +61,38 @@ public void DisableProfilingIntegration_RemovesProfilingIntegration()
...
+        _options.ProfilesSampleRate = 1.0;
+
+        var hub = GetSut();
+        var factory = (SamplingTransactionProfilerFactory)_options.TransactionProfilerFactory!;
+        Assert.False(factory.IsDisposed);
+
+        hub.Dispose();
+
+        Assert.True(factory.IsDisposed);
+    }
```
[the remark this answers is not in the document]
Comment (jamescrosswell) [reply test/Sentry.Profiling.Tests/ProfilingSentryOptionsExtensionsTests.cs:77]: It's a bit of an edge case. People usually start the SDK once per process and shut it down when the process exits... so options don't generally get reused.

We could retain a reference to the options and clear the factory if it's ours though.

@vaind thoughts?

> [!NOTE]
> This whole PR is a bit of an edge case. The 'leak' only presents when shutting down the SDK so typically when exiting the process. In practice, I think very few people will notice or be affected by this.

**Review by vaind — APPROVED** [review]
I'd look at bugbot's comment, other than that, lgtm
